### PR TITLE
nbi api for apps/appinsts

### PR DESCRIPTION
**Pull request by gainsley**
This adds support for the App/AppInstance NBI APIs, and a new Cluster API for getting cluster info.

- ClusterInsts now track the kubernetes cluster version. This can be left blank for any/unspecified version.
- Internally, we now use edgeproto.InstPort instead of dme.AppPort. InstPort is a superset of AppPort and allows for tracking more data, without having to modify the DME API, which doesn't need this extra data.
- The nbictl client allows for using object "names", which it then converts to IDs to make the actual API call. This allows for e2e tests and other clients to use infrastructure-as-code, even though the NBI API requires IDs instead of names.

### CODE REVIEW

**Review by levshvarts — APPROVED** [review]
Looks good. Thanks for adding this - so many shim structures/methods :) 
One quick question regarding internal visibility only ports, should we migrate prometheus app to use those? It seems like a better idea than having the action be implicit based on the type of application.

File: api/edgeproto/appinst.proto [hunk 1]
```diff
@@ -257,6 +257,34 @@ message AppInstRuntime {
   repeated string container_ids = 1 [(protogen.backend) = true];
 }
 
+// InstPort port information
+//
+// InstPort is a superset of dme.AppPort
+message InstPort {
+  // TCP (L4) or UDP (L4) protocol
+  distributed_match_engine.LProto proto = 1;
+  // Container port
```
[the remark this answers is not in the document]
Comment (gainsley) [reply api/edgeproto/appinst.proto:266]: Just to decouple the two APIs. That way if we need to change one or the other APIs in the future (likely to deal with CAMARA etc), then we only need to change one place in the code - where we copy from InstPort to dme.AppPort.

File: pkg/controller/appinst_zone_test.go [hunk 2]
```diff
@@ -236,6 +237,23 @@ func TestAppInstGetPotentialClusters(t *testing.T) {
...
+	app.Deployment = cloudcommon.DeploymentTypeKubernetes
+	app.AllowServerless = true
+	ai.KubernetesResources.MinKubernetesVersion = "1.30"
+	pclos, err = apis.appInstApi.getPotentialCloudlets(ctx, cctx, ai, app)
+	require.Nil(t, err)
+	verifyCloudlets(pclos, 3, 1, 4, 2, 0)
+	pclus, err = apis.appInstApi.getPotentialClusters(ctx, cctx, ai, app, pclos)
+	require.Nil(t, err)
+	for _, pc := range pclus {
+		fmt.Println(*pc)
```
[the remark this answers is not in the document]
Comment (gainsley) [reply pkg/controller/appinst_zone_test.go:252]: Yup, just for debug, will remove

File: pkg/controller/nbiconvert_app.go [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+	case nbi.VmResources:
+		return nil, fmt.Errorf("vm resources not handled yet")
+	case nbi.ContainerResources:
+		return nil, fmt.Errorf("container resources not handled yet")
+	case nbi.DockerComposeResources:
+		return nil, fmt.Errorf("docker compose resources not handled yet")
+	default:
+		return nil, fmt.Errorf("unhandled NBI App required resources type %T", rr)
+	}
+
```
[the remark this answers is not in the document]
Comment (gainsley) [reply pkg/controller/nbiconvert_app.go:149]: removed

File: pkg/controller/nbiconvert_appinst.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+	}
+	// Lookup zone ID. Note that zone may not be set if cloudlet
+	// was removed from zone.
+	zone := edgeproto.Zone{}
+	if s.allApis.zoneApi.cache.Get(&in.ZoneKey, &zone) {
+		ai.EdgeCloudZoneId = zone.ObjId
+	}
+	endpoints := []nbi.AppInstanceInfo_ComponentEndpointInfo{}
+	for _, port := range in.MappedPorts {
+		// XXX: does FqdnPrefix append or prepend?
```
[the remark this answers is not in the document]
Comment (gainsley) [reply pkg/controller/nbiconvert_appinst.go:42]: Yup, I got it backwards. It's prepend, saw it set up for a cloudlet on Azure yesterday.

**Pull request merged**